### DISCUSSION TITLE
Export Field notes to customizable Web Service

### DISCUSSION THREAD

**Discussion by baiti** [cgeo/cgeo#9544] [General]
Recently I have started to investigate opencaching.de in addition to geocaching.com, which I am using since many years.

It seems cgeo is the one and only app on the phone that supports both platforms in a nicely integrated way. Well, up to the point where it comes to "Export Field Notes".

Field Notes are exported to the local filesystem on the phone, and optionaly and ONLY for geocaching.com, they are exported to gecocaching.com's "log drafts". If I have GC* and OC* offline logs in my history, it seams uploading for OC* logs ist just ignored.

If course it is ignored, because opencaching.de does not have a OKAPI service that could be used and neither does it have any HTML pages that could be scraped....

This brings me to an idea: I would like to have a configurable URL to a webservice where the offline logs could be exported to. Regardless whether they are GC* or OC* logs.

All it takes would be a settings property do let me specify the URL and an access_token.

Could we discuss this proposal somewhere? I would be motivated to jump in for an implementation / test ;)

**Comment by Lineflyer** [comment]
AFAIR the OC team is planning to implement field note upload via OKAPI at some point in future. If this is supported, c:geo might support it of course.

**Reply by baiti** [reply]
that would absolutely help, of course. And knowing it, relieves me from the time to discuss it with them myself, however, I would like to make an additional case. With additional I mean on top of OC.de potentially supporting it.

In order to understand that case, a word on the workflow is in order. I use cgeo's field note export function in order to postprocess these logs in my own application. To avoid the "sneakernet", meaning to send files around from Smartphone (cgeo export to file) through Dropbox or Google Drive and fnally importing from a file into my application, I do the following:

I dump the logs to GC.com draft logs for the sole purpose of pulling them from there with my app. If I could put my REST API Service into a cgeo setting, together with my access_token, that woul greatly simplify my "flow graph" for both, GC.com as well as OC.de.

**Comment by baiti** [comment]
Let me add one more finding: The standard way to export field-notes from cgeo into a file on the phone, upload it to whatever storage sercvice e.g. Dropbox or Googel Drive, and import it from there as field-notes via the filesystem into opencaching.de does NOT work. Opencaching.de rejects everything I ever gave it to parse. It rejects it as as UTF-16 Big Endian, the natural format that cgeo creates, and it rejects it after conversion to a plain ASCII file as well. Most likely a bug on their side but ideally I would want to make myself independent from gc.com and oc.de as well.

From my perspective, what would be the most flexible approach is to suppert a custom URL/ access_token for uploading field-notes to whatever service the user chooses. I can provide such an endpoint for testing .. of course.

**Reply by baiti** [reply]
While I am writing this, I was wondering, has anybody ever successfully exchanged field-notes from cgeo to opencaching.* ? How?

**Comment by bekuno** [comment]
Please see the conversations at
https://redmine.opencaching.de/issues/143
https://redmine.opencaching.de/issues/1018
https://forum.opencaching.de/index.php?topic=3258

The great problem - the development for opencaching(.de) is near zero

**Reply by baiti** [reply]
After following all three references and looking at the dates when these conversations took place I'd come to the same conclusion which would make it even more important to have the service I outlined above available in cgeo.

**Comment by eddiemuc** [comment]
@baiti I wonder if what you're looking for is more some sort of generic app where other apps can send text to and this app can then forward this text to configurable URLs?

**Reply by baiti** [reply]
> @baiti I wonder if what you're looking for is more some sort of generic app where other apps can send text to and this app can then forward this text to configurable URLs?

Certainly. If such an app existed, I would use it for that purpose, however, I'd assume it would'nt make a difference for cgeo with respect to implementing the functionality to send to that magic app unless you'd assume that cgeo continues to export the file phone's filesystem and the magic app picks it up from there. It would work, of course but I would describe that approach as a kludge at best.

**Reply by fm-sys** [reply]
> It would work, of course but I would describe that approach as a kludge at best.

I'm thinking of something like we have done it for creating backups.

![Screenshot_20201216_093137_cgeo geocaching_copy_250x153](https://user-images.githubusercontent.com/64581222/102324071-98b10080-3f81-11eb-9630-d9147f0ce426.png)

Then you would be able to directly share it with your third party app, upload it to drive or whatever you like. And it would be a general solution where all users would have a benefit and it would be easy to understand... 

What do you think?

Edit: done with  #9581

**Reply by baiti** [reply]
If that existed, I believe it wouldn't simplify my daily workflow. The goal is to reduce the sequence and number of buttons to touch repeatedly on the Smartphone. Perhaps it is best if I describe the workflow in more detail.

The scenario is I am back from a caching tour and want to process the offline logs that are in cgeo. Today I go to: History -> Export -> Field Notes. In the pop up I have both currently available checkboxes checked. The "Upload to GC.com" and the "Only since ....".  Since the checkboxes stay fortunately always on, it is sufficient, when this dialog pops up to just press enter.

Once the task completes on cgeo the offline logs are on the GC.com platform and can be directly processed from there without any additional step to move the data manually.

If that task would send GC* logs to GC.com and OC* logs to OC.de my workflow would stay as compact as it is today. However, we know that OC.de's OKAPI doesn't provide the required services and most likely never will.

The best approach, best in terms of minimizing manual and repetitive button touches and selecting entries in explorer style lists on the phone would be to have an additional checkbox in cgeo's popup mentoned above:

X Upload to Geocaching.com
X Upload to <myname>
X Only since last export

I would check only the one in the middle and a single click on "Export" does the job. All logs, GC* and OC* would be POSTed to a single destination. I woulld neither export to the local filesystem nor to Geocaching.com but to my own service, which does not live on the Smartphone and consequently the "Share" function shown above could not reach that service unless another application would be written and deployed which would, if it existed, add buttons to touch, lists to select from, to the workflow which makes it worse than it is today. 

I would have an entry somewhere in global settings where I put in the following:

A name of a service
An URL
An access_token

And perhaps a checkox to enable / not enable that functionality

If these fields are configured and the checkbox set,  then the

X Upload to <myname>

would show up in the History -> Export -> Field Notes popup mentioned above.

What the task does when "Export" ist touched is simple. It POSTs the URL specified in the settings, puts the access_token into the Authorization Header and stuffs the JSON stringified field notes into the payload of the POST request. I'd take the data from there and  answer the POST with a 204.

If it is about implementing it, I would jump in and take a stab at it ... assuming that @eddiemuc would provide some advide here and there as he did  before.

A more generalized approach would be if cgeo provided a plug in architecture pretty much what Locus Map does. In that case I could go ahead and develop a plugin that does this function and probably additional functions in the future too. Knowing that this would be a **major** undertaking on cgeo's side I would opt for the solution I have outlined above.

**Reply by Lineflyer** [reply]
My personal opinion would be as @fm-sys suggested. Use the same share dialog we use for other generated files also for field notes to allow sharing with whatever a user would like. This is more generic as the approach of @baiti. Otherwise we will be forced to implement more and more export tools which should not be our focus.

**Comment by Lineflyer** [comment]
Besides the remaining discussion I did the following:
- Opened an issue for OKAPI to ask them about the field note topic: https://github.com/opencaching/okapi/issues/601
- Transfered the share menu idea into a c:geo issue: https://github.com/cgeo/cgeo/issues/9662

**Comment by Lineflyer** [comment]
Update:
As can be seen in opencaching/okapi#601 the OKAPI developers are on it.

**Reply by baiti** [reply]
Cool! .... assume for a while, they implemented the missing service, how would we integrate that into cgeo? One scenario is that the "History -> Export -> Field Notes Dialog now has three entries:

x Upload to geocaching.com
x Upload to opencaching.de
x Only since last export ....

This would leave the point open that the "Upload" checkboxes should only be presented if the platform in question is marked "active" in settings. There is a bug in the code currently as "Upload to geocaching.com" is displayed and executed, even if geocaching.com as a platform has been deactivated (but not deleted) in settings.

In addition, but I'd assume this is rather trivial: There should then be a notion of a geocache prefix (GC, OC, ....) which is used to select what  record is uploaded to what platform.

**Comment by Lineflyer** [comment]
> In addition, but I'd assume this is rather trivial: There should then be a notion of a geocache prefix (GC, OC, ....) which is used to select what record is uploaded to what platform.

Yes and no:
GC.com does expect only GC geocodes and might complain or declare the file invalid if OC are contained. But AFAIR my last test showed that it just ignores unknown geocodes.
OC does explicitly accept GC geocodes and tries to identify double listings as owners can store the GC geocode of the doubled cache in their OC listing. If there is a match OC will treat the GC cache field note as valid for that OC cache.

**Reply by baiti** [reply]
Ok in that case it seems we can present all objects at least to these two platforms and they will automatically figure out what to "ignore" (GC.com) and for what to find a match (OC.de). It will be interesting to see what OC.de does if it gets two records for the same cache, e.g. GC123 and OC456, where OC456 has GC123 "cross listed". Let's say they parse GC123 first and find out, yes, that points to our cross listed OC456, let's just create a log entry for this one and then parse the next record which presents a log for OC456 .... I would assume that is OT here and should be discussed somewhere on the OC.de side.

What remains then is that we come up with some sort of a design for that "History -> Export -> Field Notes" dialog and how exactly that should work. If I had to design / implement / test something like this, I would do a select on the object prefix (GC*, OC*) and route them to the appropriate platform based on the prefix. I am convinced that this would be the most straight forward procedure not only from a development perspective but also for the user.

**Reply by Lineflyer** [reply]
> ). It will be interesting to see what OC.de does if it gets two records for the same cache, e.g. GC123 and OC456, where OC456 has GC123 "cross listed". 

Not sure, how clever they did it. Try it out. You can already now upload such a field note file on oc.de website.
